### ISSUE TITLE
SWIG and Qt

### ISSUE THREAD

**Issue by raichu** [swig/swig#88]
Due to some Qt-specific things like "Signal/Slot, QObject parenting ownership, Qt Properties, QVariant, and custom classes for string, vectors, etc.", it's current not possible to generate full Qt bindings using SWIG.

Creating a github issue for the ideas recently discussed in [here](http://sourceforge.net/mailarchive/message.php?msg_id=30349828) and [here](http://sourceforge.net/mailarchive/message.php?msg_id=30818925) in the hopes of broader awareness and call for help.

**Comment by feddischson** [comment]
I'm very interested in contributing to support Qt. 

Was there any further progress since these discussion mentioned in the first comment? Or is there somebody working on a solution?

**Comment by ojwb** [comment]
I'm not aware of anyone working on this - you might want to ask on swig-devel though.

**Comment by gigglesun** [comment]
So it's still impossible to use SWIG to wrap QT applications?  I had see [here ](http://lists.qt-project.org/pipermail/pyside/2013-January/000957.html) seems they had successfully created a binding of Qt in Swig.

**Comment by feddischson** [comment]
I don't know a working solution, but I found some time to place my experiments here: https://github.com/feddischson/swig_qt.

It can be used to create a python extension, where you can create simple QObjects, a QMainWindow or a QPushButton. 

The meta-system is not yet working or supported. I think there are several solutions to add this functionality, but maybe somebody from the SWIG team could suggest a nice and clean solution.

In general:
I still like the idea of using SWIG for wrapping QT5. For example it would allow to create a ruby binding for QT5(at the moment, there is only http://seanchas116.github.io/ruby-qml/ which is limited, see also here: https://en.wikipedia.org/wiki/List_of_language_bindings_for_Qt_5).

At the moment, I'm not sure what the best and most efficient way for structuring such a wrapper is, but I appreciate any suggestions and hints.

**Comment by greg002** [comment]
Hi,

I try to wrap a C++ library (to C#) which use some QT Core classes:
QString, QMap, QVector, QSet

Even if i try Swig with a simple class which use QString or QVector,
I have 2 problems:
1) i do not succed to compile my ".i". I have several errors and it seems that i have to add all dependancies of QT (QChar, QTypeInfo,...)
2) How can i get a C# String from QString in my C# application, How can i set C# String to a QString ? (Same qestion for Qvector,...)

------------------------------------------------------
#include "QString"

class Qt_test
{
public:
	QString name;
	Qt_test(){ name = "test";};

	QString getName(){ return name; };
};
------------------------------------------------------------

I read the Swig documentation but i am lost.
I do not succed to understand what i need to understand and how to proceed.

I have tried to use your interface files. But it does not work for me or i don't know how to use it.
Do you plan to add a simple C++ class example with its interface file to your project ?
Even if you export to python and not C#, maybe it can help me.

Thanks
Regards,

**Comment by Ingener74** [comment]
@greg002 try this 
https://github.com/Ingener74/Lost-Foot

**Comment by greg002** [comment]
Thanks Ingener74,
but your code do not use QT. I have problems of compilating interface to wrap my lib + QString, QMap, Qvector,.... and after how to use it in C#.
I need an example (or something to help me) which export a QString.cs (QVector.cs, QMap.cs,...) which can be manipulated by C#.

Example (what it should be in C#):
```
MyClass myClass = new MyClass();
string csharp_str = new string("test");
QString str_qt = new QString(csharp_str);  // C# to QString conversion
myClass.setString(str_qt);
myClass.doSomething();
QString result = myClass.getStringResult();
csharp_str = result.toCSharpString();  // QString to C# String conversion
```

where (in C++):
`QString MYClass::getStringResult();`
-------------------------------------------------------

Example2 (what it should be in C#):
```
MyClass myClass = new MyClass();
QString str_qt = new QString(); // str_qt is empty
myClass.FillStringWithSomething(ref str_qt);     // str_qt now equal "something"
string csharp_str = str_qt.toCSharpString();
```

where (in C++):
`MYClass::FillData(QString& str);`
-------------------------------------------------------

Example 3 (what it should be in C#):
```
MyClass myClass = new MyClass();
QString strQT = new QString(); // empty
QVector vecQT = new QVector(); // empty
myClass.FillData(ref strQT, ref vecQT);
string csharp_str       =  str_qt.toCSharpString();
List<int> csharp_list = vecQT.toCSharpList();
```

where (in C++):
`MYClass::FillData(QString* str, QVector<int>& vec);`


-------------------------------------------------------

Example 4 (what it should be in C#):
```
List<int> csharp_list = new List<int>();
csharp_list.Add(1);
csharp_list.Add(2);
csharp_list.Add(3);

// C# list to QVector. Could also be vecQT.setCSharpList(csharp_list);
QVector vecQT = new QVector(csharp_list);  
myClass.setVector(vecQT);
```

where (in C++):
`MYClass::setVector(QVector vec);`

**Comment by Ingener74** [comment]
@greg002 you can copy declaration of this classes to your *.i swig interface file and swig generate wrapper for it. I try do some example for you after work.

**Comment by Ingener74** [comment]
@greg002 i hope [it](https://github.com/Ingener74/Steep-Neptune) help you. Ask if don't understand.

**Comment by wsfulton** [comment]
@greg002, FYI, I fixed the formatting in your post on 11 Aug.

**Comment by wolfgangmauer** [comment]
After a long time,,,,
Is there a working solution?

**Comment by Ingener74** [comment]
@wolfgangmauer you can saw my example, https://github.com/Ingener74/Steep-Neptune maybe it help you.

**Comment by wolfgangmauer** [comment]
Thanks for reply.
I looking for a "working" solution for Qt -> C#. (QtSharp don't work :-()

**Comment by ojwb** [comment]
I think realistically support for Qt is not going to get implemented and maintained within the SWIG project, so I'm going to close this ticket.  If someone was going to do this work, it would have happened in the 8+ years this has been open.

We have limited development resources, and there's a backlog of open tickets and unimplemented features from newer C++ standards.  Even some C++11 features are still missing, and that's over a decade old now.

So in general the policy we've ended up on is that we aim to support standard C and C++ and support for other libraries, etc is left to others (there's the odd exception where a common non-standard extension is supported, and boost shared pointers are because there's almost exactly like the standardised shared pointers).

So in this case it could be a set of Qt-specific SWIG typemaps either maintained as part of Qt, or as a separate "Qt-SWIG-typemaps" project.

I doubt any special SWIG support would be needed beyond typemaps for `QString`, `QVector`, etc (Qt's signals and slots are mentioned above, but those are implemented using macros for the compiler and those same macros should work for SWIG too).
